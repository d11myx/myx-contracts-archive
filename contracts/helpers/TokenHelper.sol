// SPDX-License-Identifier: BUSL-1.1
pragma solidity ^0.8.0;

import "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
import "@openzeppelin/contracts/utils/math/SafeMath.sol";
import "../libraries/PrecisionUtils.sol";
import "../interfaces/IPool.sol";

library TokenHelper {
    using PrecisionUtils for uint256;
    using SafeMath for uint256;

    function convertIndexAmountToStable(
        IPool.Pair memory pair,
        int256 indexTokenAmount
    ) internal view returns (int256 amount) {
        uint256 indexTokenDec = uint256(IERC20Metadata(pair.indexToken).decimals());
        uint256 stableTokenDec = uint256(IERC20Metadata(pair.stableToken).decimals());

        uint256 indexTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - indexTokenDec);
        uint256 stableTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - stableTokenDec);

        amount = (indexTokenAmount * int256(indexTokenWad)) / int256(stableTokenWad);
    }

    // function convertIndexAmountToStableWithPrice(
    //     IPool.Pair memory pair,
    //     int256 indexTokenAmount,
    //     uint256 price
    // ) internal view returns (int256 amount) {
    //     uint256 indexTokenDec = uint256(IERC20Metadata(pair.indexToken).decimals());
    //     uint256 stableTokenDec = uint256(IERC20Metadata(pair.stableToken).decimals());

    //     uint256 indexTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - indexTokenDec);
    //     uint256 stableTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - stableTokenDec);

    //     amount =
    //         ((indexTokenAmount * int256(price) * int256(indexTokenWad)) /
    //             int256(PrecisionUtils.PRICE_PRECISION)) *
    //         int256(stableTokenWad);
    // }

    function convertStableAmountToIndex(
        IPool.Pair memory pair,
        int256 stableTokenAmount
    ) internal view returns (int256 amount) {
        uint256 indexTokenDec = uint256(IERC20Metadata(pair.indexToken).decimals());
        uint256 stableTokenDec = uint256(IERC20Metadata(pair.stableToken).decimals());

        uint256 indexTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - indexTokenDec);
        uint256 stableTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - stableTokenDec);

        amount = (stableTokenAmount * int256(stableTokenWad)) / int256(indexTokenWad);
    }
}
