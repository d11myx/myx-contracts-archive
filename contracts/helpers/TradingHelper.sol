// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.0;

import '../libraries/PrecisionUtils.sol';
import '../interfaces/IAddressesProvider.sol';
import '../interfaces/IOraclePriceFeed.sol';
import '../interfaces/IIndexPriceFeed.sol';
import '../interfaces/IPool.sol';

// import 'hardhat/console.sol';

library TradingHelper {
    using PrecisionUtils for uint256;

    function getValidPrice(
        IAddressesProvider addressesProvider,
        address token,
        IPool.TradingConfig memory tradingConfig
    ) internal view returns (uint256) {
        uint256 oraclePrice = IOraclePriceFeed(addressesProvider.priceOracle()).getPrice(token);

        uint256 indexPrice = IIndexPriceFeed(addressesProvider.indexPriceOracle()).getPrice(token);

        uint256 diffP = oraclePrice > indexPrice ? oraclePrice - indexPrice : indexPrice - oraclePrice;
        diffP = diffP.calculatePercentage(oraclePrice);

        require(diffP <= tradingConfig.maxPriceDeviationP, 'exceed max price deviation');
        return oraclePrice;
    }
}
