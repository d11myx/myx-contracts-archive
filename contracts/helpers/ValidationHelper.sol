// SPDX-License-Identifier: BUSL-1.1
pragma solidity ^0.8.0;

import "../interfaces/IPositionManager.sol";
import "../interfaces/IPool.sol";
import "../interfaces/IAddressesProvider.sol";
import "../interfaces/IRoleManager.sol";
import "../libraries/PrecisionUtils.sol";
import "../libraries/TradingTypes.sol";

library ValidationHelper {
    using PrecisionUtils for uint256;

    function validateAccountBlacklist(
        IAddressesProvider addressesProvider,
        address account
    ) internal view {
        require(
            !IRoleManager(addressesProvider.roleManager()).isBlackList(account),
            "blacklist account"
        );
    }

    function validateOrderExpired(uint256 orderTime, uint256 maxTimeDelay) internal view {
        require(orderTime + maxTimeDelay >= block.timestamp, "order expired");
    }

    function validatePriceTriggered(
        IPool.TradingConfig memory tradingConfig,
        TradingTypes.TradeType tradeType,
        bool isAbove,
        uint256 currentPrice,
        uint256 orderPrice,
        uint256 maxSlippage
    ) internal pure {
        if (tradeType == TradingTypes.TradeType.MARKET) {
            bool valid = currentPrice >=
                orderPrice.mulPercentage(PrecisionUtils.percentage() - maxSlippage) &&
                currentPrice <= orderPrice.mulPercentage(PrecisionUtils.percentage() + maxSlippage);
            require(maxSlippage == 0 || valid, "exceeds max slippage");
        } else if (tradeType == TradingTypes.TradeType.LIMIT) {
            require(
                isAbove
                    ? currentPrice.mulPercentage(
                        PrecisionUtils.percentage() - tradingConfig.priceSlipP
                    ) <= orderPrice
                    : currentPrice.mulPercentage(
                        PrecisionUtils.percentage() + tradingConfig.priceSlipP
                    ) >= orderPrice,
                "not reach trigger price"
            );
        } else {
            require(
                isAbove ? currentPrice <= orderPrice : currentPrice >= orderPrice,
                "not reach trigger price"
            );
        }
    }
}
