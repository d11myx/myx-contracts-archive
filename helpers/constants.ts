import { parseUnits } from 'ethers/lib/utils';
import { BigNumber, ethers } from 'ethers';

export const ONE_ETHER = ethers.utils.parseEther('1');
export const MAX_UINT_AMOUNT = ethers.constants.MaxUint256;
export const ZERO_ADDRESS = ethers.constants.AddressZero;

export enum eEthereumNetwork {
    main = 'mainnet',
    goerli = 'goerli',
}

export enum eBscNetwork {
    main = 'mainnet',
    test = 'testnet',
}

export enum TradeType {
    MARKET = 0,
    LIMIT = 1,
    TP = 2,
    SL = 3,
}

export type eNetwork = eEthereumNetwork | eBscNetwork;

export const Market = 'USDT';

export function getMarketSymbol(): string {
    return Market;
}

export const MOCK_PRICES: { [key: string]: BigNumber } = {
    BTC: parseUnits('30000', 8),
    ETH: parseUnits('2000', 8),
};

export const DEFAULT_NAMED_ACCOUNTS = {
    deployer: {
        default: 0,
    },
    keeper: {
        default: 0,
    },
    feeReceiver: {
        default: 0,
    },
};
