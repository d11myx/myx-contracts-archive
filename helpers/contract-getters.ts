import { HardhatRuntimeEnvironment } from 'hardhat/types';
import {
    AddressesProvider,
    Executor,
    IndexPriceFeed,
    MockPriceFeed,
    OraclePriceFeed,
    Pool,
    OrderManager,
    RoleManager,
    Router,
    Token,
    PositionManager,
    WETH,
} from '../types';
import { getContract } from './utilities/tx';
import {
    ADDRESSES_PROVIDER_ID,
    INDEX_PRICE_FEED_ID,
    MOCK_PRICE_FEED_PREFIX,
    ORACLE_PRICE_FEED_ID,
    PAIR_INFO_ID,
    ROLE_MANAGER_ID,
    MOCK_TOKEN_PREFIX,
    TRADING_VAULT_ID,
    ROUTER_ID,
    EXECUTOR_ID,
    ORDER_MANAGER_ID,
    POSITION_MANAGER_ID,
} from './deploy-ids';
import { MARKET_NAME } from './env';

declare var hre: HardhatRuntimeEnvironment;

export const getMockToken = async (pair: string, address?: string): Promise<Token> => {
    return getContract<Token>('Token', address || (await hre.deployments.get(MOCK_TOKEN_PREFIX + pair)).address);
};

export const getToken = async (address?: string): Promise<Token> => {
    return getContract<Token>('Token', address || (await hre.deployments.get(MARKET_NAME)).address);
};

export const getWETH = async (address?: string): Promise<WETH> => {
    return getContract<WETH>('WETH', address || (await hre.deployments.get('WETH')).address);
};

export const getMockPriceFeed = async (pair: string, address?: string): Promise<MockPriceFeed> => {
    return getContract<MockPriceFeed>(
        'MockPriceFeed',
        address || (await hre.deployments.get(MOCK_PRICE_FEED_PREFIX + pair)).address,
    );
};

export const getAddressesProvider = async (address?: string): Promise<AddressesProvider> => {
    return getContract<AddressesProvider>(
        'AddressesProvider',
        address || (await hre.deployments.get(ADDRESSES_PROVIDER_ID)).address,
    );
};

export const getRoleManager = async (address?: string): Promise<RoleManager> => {
    return getContract<RoleManager>('RoleManager', address || (await hre.deployments.get(ROLE_MANAGER_ID)).address);
};

export const getOraclePriceFeed = async (address?: string): Promise<OraclePriceFeed> => {
    return getContract<OraclePriceFeed>(
        'OraclePriceFeed',
        address || (await hre.deployments.get(ORACLE_PRICE_FEED_ID)).address,
    );
};

export const getIndexPriceFeed = async (address?: string): Promise<IndexPriceFeed> => {
    return getContract<IndexPriceFeed>(
        'IndexPriceFeed',
        address || (await hre.deployments.get(INDEX_PRICE_FEED_ID)).address,
    );
};

export const getPool = async (address?: string): Promise<Pool> => {
    return getContract<Pool>('Pool', address || (await hre.deployments.get(PAIR_INFO_ID)).address);
};

export const getRouter = async (address?: string): Promise<Router> => {
    return getContract<Router>('Router', address || (await hre.deployments.get(ROUTER_ID)).address);
};

export const getExecutor = async (address?: string): Promise<Executor> => {
    return getContract<Executor>('Executor', address || (await hre.deployments.get(EXECUTOR_ID)).address);
};

export const getOrderManager = async (address?: string): Promise<OrderManager> => {
    return getContract<OrderManager>('OrderManager', address || (await hre.deployments.get(ORDER_MANAGER_ID)).address);
};

export const getPositionManager = async (address?: string): Promise<PositionManager> => {
    return getContract<PositionManager>(
        'PositionManager',
        address || (await hre.deployments.get(POSITION_MANAGER_ID)).address,
    );
};
