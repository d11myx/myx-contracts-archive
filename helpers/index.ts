export * from './constants';
export * from './contract-deployments';
export * from './contract-getters';
export * from './deploy-ids';
export * from './env';
export * from './hardhat-config-helpers';
export * from './init-helper';
export * from './market-config-helper';
export * from './types';
export * from './utilities/signer';
export * from './utilities/tx';
export * from '../tasks/misc/print-deployments';

import { loadTasks } from './hardhat-config-helpers';

const TASK_FOLDERS = ['../tasks/misc'];

// Load all plugin tasks
loadTasks(TASK_FOLDERS);
