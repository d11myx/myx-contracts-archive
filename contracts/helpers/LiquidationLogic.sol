// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.20;

import "../libraries/Position.sol";
import "../interfaces/IPool.sol";
import "../interfaces/IOrderManager.sol";
import "../interfaces/IPositionManager.sol";
import "../interfaces/IAddressesProvider.sol";
import "../helpers/TradingHelper.sol";
import "../interfaces/IExecutor.sol";

library LiquidationLogic {
    using PrecisionUtils for uint256;
    using Position for Position.Info;

    uint256 public constant MAX_RATIO = 3e6;

    event ExecuteLiquidation(
        bytes32 positionKey,
        address account,
        uint256 pairIndex,
        bool isLong,
        uint256 collateral,
        uint256 sizeAmount,
        uint256 price
    );

    function liquidationPosition(
        IPool pool,
        IOrderManager orderManager,
        IPositionManager positionManager,
        IExecutor executor,
        IAddressesProvider addressesProvider,
        bytes32 positionKey,
        uint8 level,
        uint256 commissionRatio
    ) external {
        require(commissionRatio < MAX_RATIO, "max ratio");
        Position.Info memory position = positionManager.getPositionByKey(positionKey);
        if (position.positionAmount == 0) {
            return;
        }
        IPool.Pair memory pair = pool.getPair(position.pairIndex);
        IPool.TradingConfig memory tradingConfig = pool.getTradingConfig(position.pairIndex);

        uint256 price = TradingHelper.getValidPrice(
            addressesProvider,
            pair.indexToken,
            tradingConfig
        );

        int256 unrealizedPnl = position.getUnrealizedPnl(position.positionAmount, price);
        uint256 tradingFee = positionManager.getTradingFee(
            position.pairIndex,
            position.isLong,
            position.positionAmount
        );
        int256 fundingFee = positionManager.getFundingFee(
            position.account,
            position.pairIndex,
            position.isLong
        );
        int256 exposureAsset = int256(position.collateral) +
            unrealizedPnl -
            int256(tradingFee) +
            fundingFee;

        bool needLiquidate;
        if (exposureAsset <= 0) {
            needLiquidate = true;
        } else {
            uint256 riskRate = position
                .positionAmount
                .mulPrice(position.averagePrice)
                .mulPercentage(tradingConfig.maintainMarginRate)
                .calculatePercentage(uint256(exposureAsset));
            needLiquidate = riskRate >= PrecisionUtils.percentage();
        }
        if (!needLiquidate) {
            return;
        }

        // cancel all positionOrders
        orderManager.cancelAllPositionOrders(position.account, position.pairIndex, position.isLong);

        uint256 orderId = orderManager.createOrder(
            TradingTypes.CreateOrderRequest({
                account: position.account,
                pairIndex: position.pairIndex,
                tradeType: TradingTypes.TradeType.MARKET,
                collateral: 0,
                openPrice: price,
                isLong: position.isLong,
                sizeAmount: -int256(position.positionAmount),
                maxSlippage: 0,
                data: abi.encode(position.account)
            })
        );

        executor.executeDecreaseOrder(
            orderId,
            TradingTypes.TradeType.MARKET,
            level,
            commissionRatio
        );

        emit ExecuteLiquidation(
            positionKey,
            position.account,
            position.pairIndex,
            position.isLong,
            position.collateral,
            position.positionAmount,
            price
        );
    }
}
