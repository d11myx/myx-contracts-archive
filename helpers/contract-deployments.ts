import { HardhatRuntimeEnvironment } from 'hardhat/types';
import {
    AddressesProvider,
    ExecutionLogic,
    Executor,
    FeeCollector,
    FundingRate,
    IndexPriceFeed,
    LiquidationLogic,
    MockPyth,
    PythOraclePriceFeed,
    OrderManager,
    Pool,
    PoolTokenFactory,
    PositionManager,
    RiskReserve,
    RoleManager,
    Router,
    TestCallBack,
    Token,
    WETH,
    Timelock,
} from '../types';
import { Contract, ethers } from 'ethers';
import { MARKET_NAME } from './env';
import { Duration, deployContract, deployUpgradeableContract, encodeParameterArray, increase, latest, waitForTx } from './utilities/tx';
import { MOCK_INDEX_PRICES, MOCK_PRICES } from './constants';
import { SymbolMap } from './types';
import { SignerWithAddress } from '../test/helpers/make-suite';
import { loadReserveConfig } from './market-config-helper';
import { getWETH } from './contract-getters';
import { POSITION_MANAGER_ID } from './deploy-ids';
import usdt from '../markets/usdt';

declare var hre: HardhatRuntimeEnvironment;

export const deployMockToken = async (symbol: string): Promise<Token> => {
    return await deployContract<Token>('Token', [symbol]);
};

export const deployWETH = async (): Promise<WETH> => {
    return await deployContract<WETH>('WETH', ['WETH', 'WETH', '18']);
};

const logFlag = false;

export function log(message?: any, ...optionalParams: any[]) {
    if (logFlag) {
        console.log(message, ...optionalParams);
    }
}

export async function deployLibraries() {
    log(` - setup libraries`);

    const validationHelper = await deployContract('ValidationHelper', []);
    log(`deployed ValidationHelper at ${validationHelper.address}`);

    return {
        validationHelper,
    };
}

export async function deployToken() {
    log(` - setup tokens`);

    // basic token
    const usdt = await deployMockToken(MARKET_NAME);
    log(`deployed USDT at ${usdt.address}`);

    const weth = await deployWETH();
    log(`deployed WETH at ${weth.address}`);

    // pairs token
    const pairConfigs = loadReserveConfig(MARKET_NAME)?.PairsConfig;

    const tokens: SymbolMap<Token> = {};
    for (let pair of Object.keys(pairConfigs)) {
        const token = await deployMockToken(pair);
        log(`deployed ${pair} at ${token.address}`);

        tokens[pair] = token;
    }
    return { usdt, weth, tokens };
}

export async function deployPrice(
    deployer: SignerWithAddress,
    keeper: SignerWithAddress,
    timelock: Timelock,
    addressesProvider: AddressesProvider,
    tokens: SymbolMap<Token>,
) {
    log(` - setup price`);

    const mockPyth = (await deployContract('MockPyth', [60, 1])) as any as MockPyth;

    const oraclePriceFeed = (await deployContract('PythOraclePriceFeed', [
        addressesProvider.address,
        mockPyth.address,
        [],
        [],
    ])) as any as PythOraclePriceFeed;
    log(`deployed PythOraclePriceFeed at ${oraclePriceFeed.address}`);

    const pairTokenAddresses = [];
    const pairTokenPrices = [];
    const pairTokenIndexPrices = [];
    const pairTokenPriceIds = [];
    for (let [pair, token] of Object.entries(tokens)) {
        const pairTokenAddress = token.address;
        if (!pairTokenAddress) {
            throw `wait for deployed before using`;
        }

        pairTokenAddresses.push(pairTokenAddress);
        pairTokenPrices.push(MOCK_PRICES[pair]);
        pairTokenIndexPrices.push(MOCK_INDEX_PRICES[pair]);
        pairTokenPriceIds.push(ethers.utils.formatBytes32String(pair));
    }

    const indexPriceFeed = (await deployContract('IndexPriceFeed', [
        addressesProvider.address,
        [],
        [],
    ])) as any as IndexPriceFeed;
    log(`deployed IndexPriceFeed at ${indexPriceFeed.address}`);

    await indexPriceFeed.connect(keeper.signer).updatePrice(pairTokenAddresses, pairTokenIndexPrices);

    let timestamp = await latest();
    let eta = Duration.days(1);
    await timelock.queueTransaction(
        oraclePriceFeed.address,
        0,
        'setAssetPriceIds(address[],bytes32[])',
        encodeParameterArray(['address[]', 'bytes32[]'], [pairTokenAddresses, pairTokenPriceIds]),
        eta.add(timestamp),
    );
    await increase(Duration.days(1));
    await waitForTx(
        await timelock.executeTransaction(
            oraclePriceFeed.address,
            0,
            'setAssetPriceIds(address[],bytes32[])',
            encodeParameterArray(['address[]', 'bytes32[]'], [pairTokenAddresses, pairTokenPriceIds]),
            eta.add(timestamp),
        ),
    );
    const updateData = await oraclePriceFeed.getUpdateData(pairTokenAddresses, pairTokenPrices);
    const fee = mockPyth.getUpdateFee(updateData);
    await oraclePriceFeed.connect(keeper.signer).updatePrice(pairTokenAddresses, pairTokenPrices, { value: fee });

    const fundingRate = (await deployUpgradeableContract('FundingRate', [
        addressesProvider.address,
    ])) as any as FundingRate;
    log(`deployed FundingRate at ${fundingRate.address}`);

    await addressesProvider
        .connect(deployer.signer)
        .initialize(oraclePriceFeed.address, indexPriceFeed.address, fundingRate.address);
    return { oraclePriceFeed, indexPriceFeed, fundingRate };
}

export async function deployPair(
    addressProvider: AddressesProvider,
    vaultPriceFeed: PythOraclePriceFeed,
    deployer: SignerWithAddress,
    weth: WETH,
) {
    log(` - setup pairs`);
    const poolTokenFactory = (await deployContract('PoolTokenFactory', [addressProvider.address])) as PoolTokenFactory;
    const pool = (await deployUpgradeableContract('Pool', [
        addressProvider.address,
        poolTokenFactory.address,
    ])) as any as Pool;
    log(`deployed Pool at ${pool.address}`);

    //TODO uniswap config
    // await pool.setRouter(ZERO_ADDRESS);
    // await pool.updateTokenPath();

    return { poolTokenFactory, pool };
}

export async function deployTrading(
    deployer: SignerWithAddress,
    poolAdmin: SignerWithAddress,
    addressProvider: AddressesProvider,
    roleManager: RoleManager,
    pool: Pool,
    pledge: Token,
    validationHelper: Contract,
) {
    log(` - setup trading`);

    const weth = await getWETH();
    // const usdt = await getToken();

    let feeCollector = (await deployUpgradeableContract('FeeCollector', [
        addressProvider.address,
        pool.address,
        pledge.address,
    ])) as any as FeeCollector;
    let riskReserve = (await deployUpgradeableContract('RiskReserve', [
        deployer.address,
        addressProvider.address,
    ])) as any as RiskReserve;

    let positionManager = (await deployUpgradeableContract('PositionManager', [
        addressProvider.address,
        pool.address,
        pledge.address,
        feeCollector.address,
        riskReserve.address,
    ])) as any as PositionManager;
    log(`deployed PositionManager at ${positionManager.address}`);

    let orderManager = (await deployUpgradeableContract('OrderManager', [
        addressProvider.address,
        pool.address,
        positionManager.address,
    ])) as any as OrderManager;
    log(`deployed OrderManager at ${orderManager.address}`);

    let router = (await deployContract('Router', [
        weth.address,
        addressProvider.address,
        orderManager.address,
        pool.address,
    ])) as Router;
    log(`deployed Router at ${router.address}`);
    await waitForTx(await orderManager.setRouter(router.address));

    let liquidationLogic = (await deployContract('LiquidationLogic', [
        addressProvider.address,
        pool.address,
        orderManager.address,
        positionManager.address,
    ])) as any as LiquidationLogic;
    log(`deployed LiquidationLogic at ${liquidationLogic.address}`);

    let executionLogic = (await deployContract('ExecutionLogic', [
        addressProvider.address,
        pool.address,
        orderManager.address,
        positionManager.address,
        feeCollector.address,
        60 * 10, //todo testing time
    ])) as any as ExecutionLogic;
    log(`deployed ExecutionLogic at ${executionLogic.address}`);

    let executor = (await deployUpgradeableContract('Executor', [
        addressProvider.address,
        executionLogic.address,
        liquidationLogic.address,
    ])) as any as Executor;
    log(`deployed Executor at ${executor.address}`);
    log(`executionLogic pool : ${await executor.executionLogic()}`);

    await waitForTx(await feeCollector.updatePositionManagerAddress(positionManager.address));

    await waitForTx(await pool.connect(poolAdmin.signer).setRiskReserve(riskReserve.address));
    await waitForTx(await pool.connect(poolAdmin.signer).setFeeCollector(feeCollector.address));

    await waitForTx(await riskReserve.connect(poolAdmin.signer).updatePositionManagerAddress(positionManager.address));
    await waitForTx(await riskReserve.connect(poolAdmin.signer).updatePoolAddress(pool.address));

    await waitForTx(await executionLogic.connect(poolAdmin.signer).updateExecutor(executor.address));
    await waitForTx(await liquidationLogic.connect(poolAdmin.signer).updateExecutor(executor.address));

    await positionManager.updateExecutionLogic(executionLogic.address);
    await positionManager.updateLiquidationLogic(liquidationLogic.address);
    // await positionManager.addLogic(orderManager.address);
    await orderManager.setExecutionLogic(executionLogic.address);
    await orderManager.setLiquidationLogic(liquidationLogic.address);

    return {
        positionManager,
        router,
        executionLogic,
        liquidationLogic,
        executor,
        orderManager,
        riskReserve,
        feeCollector,
    };
}
export async function deployMockCallback() {
    return (await deployContract('TestCallBack', [])) as TestCallBack;
}
