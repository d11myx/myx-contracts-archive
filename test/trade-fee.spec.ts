import { testEnv } from './helpers/make-suite';
import { ethers } from 'hardhat';
import { expect } from './shared/expect';
import { getPositionTradingFee } from '../helpers';

describe('Trade: FeeCal', () => {
    const pairIndex = 1;
    // before('add liquidity', async () => {
    //     testEnv = await newTestEnv();
    //     const {
    //         users: [depositor],
    //         usdt,
    //         btc,
    //         pool,
    //         router,
    //     } = testEnv;
    //
    //     // add liquidity
    //     const indexAmount = ethers.utils.parseUnits('10', 18);
    //     const stableAmount = ethers.utils.parseUnits('300000', 18);
    //     const pair = await pool.getPair(pairIndex);
    //     await mintAndApprove(testEnv, btc, indexAmount, depositor, router.address);
    //     await mintAndApprove(testEnv, usdt, stableAmount, depositor, router.address);
    //
    //     await router
    //         .connect(depositor.signer)
    //         .addLiquidity(pair.indexToken, pair.stableToken, indexAmount, stableAmount);
    // });

    it('calculate trading fee', async () => {
        const { positionManager, btc, usdt, oraclePriceFeed } = testEnv;

        const long = await positionManager.getTradingFee(
            pairIndex,
            true,
            ethers.utils.parseUnits('100', await btc.decimals()),
            await oraclePriceFeed.getPrice(btc.address),
        );

        let positionTradingFee = await getPositionTradingFee(
            testEnv,
            pairIndex,
            btc,
            usdt,

            ethers.utils.parseUnits('100', await btc.decimals()),
            true,
        );
        expect(long).to.be.eq(positionTradingFee);

        const short = await positionManager.getTradingFee(
            pairIndex,
            false,
            ethers.utils.parseUnits('100', await btc.decimals()),
            await oraclePriceFeed.getPrice(btc.address),
        );
        positionTradingFee = await getPositionTradingFee(
            testEnv,
            pairIndex,
            btc,
            usdt,
            ethers.utils.parseUnits('100', await btc.decimals()),
            false,
        );
        expect(short).to.be.eq(positionTradingFee);
    });
});
