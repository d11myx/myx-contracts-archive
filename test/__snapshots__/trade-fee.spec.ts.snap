// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trade: FeeCal calculate trading fee 1`] = `314170`;
